### PR TITLE
Find common errors in translations

### PR DESCRIPTION
**Pull request by hacodeorg**
[FND-1955](https://codedotorg.atlassian.net/browse/FND-1955): Create a module containing functions to find the common errors in translations. The first 3 functions are:
- Validating markdown link `[text](url)`
- Validating redacted blocks `[name][number]`
- Validating language of a translation.

## Links
[Technical proposal to reduce common translation errors](https://docs.google.com/document/d/1mbc1IaQVo708pc3AHPtK1KhrM43xhfbbtFUhbT_-TWc/edit)

## Testing story
- Unit tests: `bundle exec ruby -Ilib:test bin/i18n/test/translation_monitor/test_translation_validator.rb`

## PR Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: bin/i18n/translation_monitor/translation_validator.rb [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+    str.match?(/\[\w*\]\s+\(\w*\)/) ? 'cannot have space between [] and () block' : nil
+  end
+
+  # Redacted string should not have space(s) between two [] blocks.
+  # E.g. "[A] [0]" is invalid but "[A][0] [B][1]" is fine.
+  # @param str [String]
+  # @return [String, nil] an error message or nil if the input is valid.
+  def validate_redacted_blocks(str)
+    # delte all valid redacted blocks
+    valid_blocks_pattern = /\[\w+\]\[\w+\]/
```
Comment (annaxuphoto) [inline bin/i18n/translation_monitor/translation_validator.rb]: I think the 2nd bracket group should be numbers only in order to be a valid redaction.

File: bin/i18n/translation_monitor/translation_validator.rb [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+    remainder = str.gsub(valid_blocks_pattern, '')
```
Comment (annaxuphoto) [inline bin/i18n/translation_monitor/translation_validator.rb:28]: why are we deleting valid redactions? I'm assuming seeing a valid redaction in the output means we had a failed restoration - will this change make those failures harder to catch?

File: bin/i18n/translation_monitor/translation_validator.rb [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+
+    invalid_blocks_pattern = /\[\w+\]\s+\[\w+\]/
```
Comment (annaxuphoto) [inline bin/i18n/translation_monitor/translation_validator.rb]: Same numbers comment as above

File: bin/i18n/translation_monitor/translation_validator.rb [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+  #   is not in the expected language or if the library could not detect language reliably.
+  # @see The supported language codes at https://github.com/google/cld3/blob/master/README.md#supported-languages
+  def validate_language(str, language_code)
+    minimum_string_length = 20
+    ignored_language_codes = %w[un en]
+
+    return if str.size < minimum_string_length
+    detected_language = CLD.detect_language(str)[:code]
+    return if ignored_language_codes.include?(detected_language) || language_code == detected_language
+    "language looks like '#{detected_language}' instead of '#{language_code}'"
```
Comment (annaxuphoto) [inline bin/i18n/translation_monitor/translation_validator.rb:52]: cool! do we need minimum 20 chars to detect language?

**Review by annaxuphoto — COMMENTED** [review]
couple short qs, thanks for doing this Ha!!

File: bin/i18n/translation_monitor/translation_validator.rb (hunk 4 above)
Comment (hacodeorg) [reply bin/i18n/translation_monitor/translation_validator.rb:52]: yes, language detection is not accurate for short strings. I'll add a comment.

File: bin/i18n/translation_monitor/translation_validator.rb (hunk 1 above)
Comment (hacodeorg) [reply bin/i18n/translation_monitor/translation_validator.rb]: Good point, will fix

File: bin/i18n/translation_monitor/translation_validator.rb (hunk 2 above)
Comment (hacodeorg) [reply bin/i18n/translation_monitor/translation_validator.rb:28]: This is before restoration; the function just checks the raw translations. Will add a comment.

Comment (hacodeorg) [reply bin/i18n/translation_monitor/translation_validator.rb:28]: The logic here is to ignore all valid redacted blocks first. If we can still find things that looks like `[something]spaces[something]` after that, they must be invalid.

File: bin/i18n/test/translation_monitor/test_translation_validator.rb [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+
+    # invalid cases
+    refute_nil validate_markdown_link('[text] (url)')
+    refute_nil validate_markdown_link("[text]\t(url)")
+  end
+
+  def test_validate_redacted_blocks
+    # valid cases
+    assert_nil validate_redacted_blocks('[A][0]')
+    assert_nil validate_redacted_blocks('[A][0] [B][1]')
```
Comment (smusoke) [inline bin/i18n/test/translation_monitor/test_translation_validator.rb:21]: Do we ever have no space between two redacted blocks? If so is that accounted for?

Comment (hacodeorg) [reply bin/i18n/test/translation_monitor/test_translation_validator.rb:21]: good idea, I'll add that test

**Review by smusoke — APPROVED** [review]
👍

**Pull request merged**